src/io/fs/memory.rs

```diff
@@ -0,0 +1,685 @@
+        Self {
+            root: Arc::new(RwLock::new(Directory::new())),
+        }
+    }
+
+    /// Load the contents of a file on the local file system into the memory file system.
+    pub fn load_from_disk(
+        &self,
+        from_disk: impl AsRef<Path>,
+        to_internal: impl AsRef<Path>,
+    ) -> io::Result<()> {
+        let bytes = std::fs::read(from_disk)?;
+        let mut writer = self.create(to_internal)?;
+        writer.write_all(&bytes)?;
+        Ok(())
+    }
+    /// Write the contents of a  file in the memory file system to the local file system.
+    pub fn save_to_disk(
+        &self,
+        from_internal: impl AsRef<Path>,
+        to_external: impl AsRef<Path>,
+    ) -> io::Result<()> {
+        let mut reader = io::BufReader::new(self.open(from_internal)?);
+        let mut writer = io::BufWriter::new(std::fs::File::create(to_external)?);
+        std::io::copy(&mut reader, &mut writer)?;
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct Directory {
+    subdirs: HashMap<String, Directory>,
+    files: HashMap<String, FileEntry>,
+}
+
+impl Directory {
+    /// Create a new empty directory.
+    fn new() -> Self {
+        Self {
+            subdirs: HashMap::default(),
+            files: HashMap::default(),
+        }
+    }
+
+    fn get_directory(&self, path: impl AsRef<Path>) -> Result<&Directory, io::Error> {
+        let path = path.as_ref();
+        let mut dir: &Directory = self;
+        for name in &resolve_path(path)? {
+            dir = match dir.subdirs.get(name) {
+                Some(subdir) => subdir,
+                None => {
+                    return Err(io::Error::new(
+                        io::ErrorKind::NotFound,
+                        "directory not found",
+                    ))
+                }
+            };
+        }
+        Ok(dir)
+    }
+
+    fn get_directory_mut(&mut self, path: impl AsRef<Path>) -> Result<&mut Directory, io::Error> {
+        let path = path.as_ref();
+        let mut dir: &mut Directory = self;
+        for name in &resolve_path(path)? {
+            dir = match dir.subdirs.get_mut(name) {
+                Some(subdir) => subdir,
+                None => {
+                    return Err(io::Error::new(
+                        io::ErrorKind::NotFound,
+                        "directory not found",
+                    ))
+                }
+            };
+        }
+        Ok(dir)
+    }
+}
+
+/// Resolve a path to a canonical path (removing "..", "." and "/") containing only the direct path coponents.
+fn resolve_path(path: &Path) -> Result<Vec<String>, io::Error> {
+    let mut part: Vec<String> = Vec::new();
+    for component in path.components() {
+        match component {
+            Component::Normal(component) => {
+                let name = component.to_string_lossy().to_string();
+
+                part.push(name);
+            }
+            Component::ParentDir => {
+                if part.pop().is_none() {
```

Comment:
Should this take into account the current directory rather than just what is in `part`? If I understand correctly, parsing `../b` would fail even if the current working dir is `/a`

Reply:
Well in this implementation there is no notion of a "working directory", and paths that are resolved are always relative to the "root". The `Directory` instances do not have a link back to their parent, and the `get_directory` is only ever called on the root `Directory` instance. Thus a path like `../a` does not make sense (does not exist). Perhaps we should move those `get_` functions (and `resolve_path`) directly into `MemoryFileSystem` and not keep them in `Directory` to make it clearer, and also add a unit test 🚀

Reply:
@carlogico Improved this in 7458fedab8b536966db9f5bdea3981a34db98f93. What do you think?

Reply:
Thanks for the explanation! And the new commit does make that pretty explicit.